torch/distributed/fsdp/_runtime_utils.py

```diff
@@ -770,14 +781,16 @@ def _post_backward_hook(
                 grad_to_offload = _reduce_grad(state, handle)
             else:
                 grad_to_offload = _reduce_grad_no_shard(state, handle)
-            _offload_grad(state, handle, grad_to_offload)
+            already_ran_post_reduction_logic = grad_to_offload is None
+            if not already_ran_post_reduction_logic:
+                _offload_grad(state, handle, grad_to_offload)
+                _post_backward_use_sharded_grad_views(handle)
```

Comment:
Maybe some comment saying that this has to be synchronized with the repeated logic in `_reduce_grad`